scripts/Stretch.py

```diff
@@ -0,0 +1,419 @@
+                   + "mm and stretch=" + str(self.stretch)+ "mm")
+        retdata = []
+        layer_steps = []
+        current = GCodeStep(0)
+        layer_z = 0.
+        for layer in data:
+            lines = layer.rstrip("\n").split("\n")
+            for line in lines:
+                current.comment = ""
+                if line.find(";") >= 0:
+                    current.comment = line[line.find(";"):]
+                if _getValue(line, "G") == 0:
+                    current.readStep(line)
+                    onestep = GCodeStep(0)
+                    onestep.copyPosFrom(current)
+                elif _getValue(line, "G") == 1:
+                    current.readStep(line)
+                    onestep = GCodeStep(1)
+                    onestep.copyPosFrom(current)
+                elif _getValue(line, "G") == 92:
+                    current.readStep(line)
+                    onestep = GCodeStep(-1)
+                    onestep.copyPosFrom(current)
+                else:
+                    onestep = GCodeStep(-1)
+                    onestep.copyPosFrom(current)
+                    onestep.comment = line
+                if current.step_z != layer_z:
+                    Logger.log("d", "Layer Z " + "{:.3f}".format(layer_z)
+                               + " " + str(len(layer_steps)) + " steps")
+                    if len(layer_steps):
+                        retdata.append(self.processLayer(layer_steps))
+                    layer_steps = []
+                    layer_z = current.step_z
+                layer_steps.append(onestep)
+        if len(layer_steps):
+            retdata.append(self.processLayer(layer_steps))
+        retdata.append(";Stretch distance " + str(self.stretch) + "\n")
+        return retdata
+
+    def processLayer(self, layer_steps):
+        """
+        Computes the new coordinates of g-code steps
+        for one layer (all the steps at the same Z coordinate)
+        """
+        layergcode = ""
+        self.vd1 = np.empty((0, 2))
+        self.vd2 = np.empty((0, 2))
+        current_e = layer_steps[0].step_e
+        orig_seq = np.empty((0, 2))
+        iflush = 0
+        for i, step in enumerate(layer_steps):
+            if i == 0:
+                current_e = step.step_e
+            if current_e == step.step_e:
+                # No extrusion since the previous step, so it is a travel move
+                # Let process steps accumulated into orig_steps,
+                # which are a sequence of continuous extrusion
+                modif_seq = np.copy(orig_seq)
+                if len(orig_seq) >= 2:
+                    self.workOnSequence(orig_seq, modif_seq)
+                layergcode = self.generate(layer_steps, iflush, i, modif_seq, layergcode)
+                iflush = i
+                orig_seq = np.empty((0, 2))
+            if step.step == 0 or step.step == 1:
+                orig_seq = np.concatenate([orig_seq, np.array([[step.step_x, step.step_y]])])
```

Comment:
I have no clue to make this algorithm faster. It is necessary to know before each extrusion move if material was already deposited at the same Z position (same layer) near the new deposited segment. When drawing on a bitmap, this is very easy: Is there colored pixels around the position where we are going to draw. It is a constant time algorithm.
The problem with the bitmap approach is the amount of memory needed. To have good precision, pixels should be half of the line width. Let say 0,2mm wide.
For a maximum plate dimensions of 20cm x 20cm (and as you already mentioned, some printer are bigger), we need and array of 1000 x 1000 pixels.
Well. It seems doable. Will think of it...